### PR TITLE
Bump borsh to 0.10.3

### PR DESCRIPTION
**Pull request by kevinji**
#### Problem
borsh is on 0.9.3 right now.

#### Summary of Changes

Bumps all uses of borsh to 0.10.3. The only breaking change is to the interface of `BorshDeserialize`, which was updated in one spot, but otherwise the serialization remains the same.

transaction-status relies on SPL which still requires borsh 0.9, so until SPL also gets updated that package alone will use an older version of borsh.

Fixes #

### CODE REVIEW

File: transaction-status/Cargo.toml [hunk 1]
```diff
@@ -13,7 +13,7 @@ edition = { workspace = true }
 Inflector = { workspace = true }
 base64 = { workspace = true }
 bincode = { workspace = true }
-borsh = { workspace = true }
+borsh0-9 = { package = "borsh", version = "0.9.3" }
```
Comment (joncinque) [inline transaction-status/Cargo.toml:16]: nit: Can you add a comment to use the workspace-level version once `spl-associated-token-account` uses borsh 0.10?

**Review by joncinque — COMMENTED** [review]
The changes look great, thanks for doing this! Just one tiny nit to address.

There's also currently an error in CI because the Cargo.lock changes haven't been propagated to the other lockfile in `programs/sbf`.

To update the other lockfile, you'll need to run `./scripts/cargo-for-all-lock-files.sh tree` from the top-level of the repo and then commit your changes.

I ran clippy, fmt, and the tests against this locally, so it should pass after that. Ping me here once you've updated the branch so I can kick off CI!

**Comment by kevinji** [comment]
@joncinque Just pushed a commit addressing your comments.

**Comment by kevinji** [comment]
The build failures seem to be because of SPL compile issues which would be expected for now until SPL also switches over to borsh 0.10. Is there anything that should be done here?

**Comment by joncinque** [comment]
@kevinji we had a discussion about this internally. When we upgrade borsh, we need to disable the downstream CI jobs because of the new incompatibility, e.g. https://github.com/solana-labs/solana/pull/18218#issuecomment-868726180 -- 

To fix that, you'll need to comment out the lines for spl and openbook-dex and reorder at https://github.com/solana-labs/solana/blob/master/.buildkite/scripts/build-downstream-projects.sh

Since we'll need to disable CI, we'd like to hold off merging this until we're about to branch 1.16.0. The timeline for that isn't clear unfortunately, so we may need to keep this open for some time.

**Comment by kevinji** [comment]
Makes sense; feel free to leave this PR alone until that happens.

**Comment by joncinque** [comment]
Thanks so much for taking care of this so quickly and for the great work!

**Comment by joncinque** [comment]
It looks like 1.16 branch cut is imminent so we should rebase and land this asap -- do you have time to do that @kevinji ? If not, I can take care of it.

Any objections about that @CriesofCarrots ?

**Comment by kevinji** [comment]
I don’t currently have the bandwidth for this so if you could take care of the rebase that would be great!

**Comment by joncinque** [comment]
@willhickey this one

**Comment by joncinque** [comment]
@kevinji thanks for being so responsive! I went ahead and rebased. @CriesofCarrots can you take one more look?

**Review by CriesofCarrots — APPROVED** [review]

**Comment by acheroncrypto** [comment]
This bump has caused massive amounts of dependency issues for all programs that depend on other versions of `borsh`.

Not only this has broken all older program dependencies, it has also broken publishing to `crates.io` when you have various dependencies with different `borsh` versions.

Even with the same Solana version(`1.16.0`) crates differ on which `borsh` version they use, e.g. `solana-transaction-status` is still on `0.9.3` when others are not.

**Comment by joncinque** [comment]
It might be worth creating a separate issue to discuss this -- do you have an example setup that shows the issue? `solana-transaction-status` uses both versions https://github.com/solana-labs/solana/blob/37f51e8376248ab3c8fc26f4ca8483fbada37da1/transaction-status/Cargo.toml#L16-L17 to get around the issue for example.

It might be that certain projects can't upgrade to 1.16 until all of the other dependencies do too, or they might need to use 1.14 and 1.16 concurrently to get around dep issues.

We've had these bumps before, and it's certainly annoying and takes time (see https://github.com/solana-labs/solana/pull/18218) but it's doable

**Comment by KirillLykov** [comment]
@joncinque this PR disabled the spl and opendex downstream builds, are we ready to re-enable them and if not, are you aware if we have an issue to track this (if not I will create one)?

**Comment by joncinque** [comment]
Correct, we can re-enable them, but we'll have to add a fix to maintain borsh's transitive dependency on hashbrown to 0.12.3, since cargo will automatically upgrade it during the downstream build, causing the program builds to fail.

**Comment by KirillLykov** [comment]
> Correct, we can re-enable them, but we'll have to add a fix to maintain borsh's transitive dependency on hashbrown to 0.12.3, since cargo will automatically upgrade it during the downstream build, causing the program builds to fail.

I've created an issue to track this https://github.com/solana-labs/solana/issues/32378

**Pull request merged**